oodeel/methods/dknn.py

```diff
@@ -56,54 +69,174 @@ def __init__(self, nearest: int = 50, use_gpu: bool = False):
+            using the computed scores.
+
+        Args:
+            fit_dataset (Union[TensorType, DatasetType]): In-distribution data used for
+                constructing the indices.
+        """
+        self.indexes = []
+        num_feature_layers = len(self.feature_extractor.feature_layers_id)
+
+        if self.postproc_fns is None:
+            self.postproc_fns = [
+                self.feature_extractor._default_postproc_fn
+            ] * num_feature_layers
+
+        fit_projected, _ = self.feature_extractor.predict(
+            fit_dataset, postproc_fns=self.postproc_fns
+        )
+        per_layer_scores = []
+
+        # If there is more than one feature layer, ensure an aggregator is defined.
+        if self.aggregator is None and num_feature_layers > 1:
+            self.aggregator = StdNormalizedAggregator()
+
+        for i in range(num_feature_layers):
+            index, layer_scores = self._fit_layer(fit_projected[i])
+            self.indexes.append(index)
+            if self.aggregator is not None:
+                per_layer_scores.append(layer_scores)
+
+        if self.aggregator is not None:
+            self.aggregator.fit(per_layer_scores)
 
     def _score_tensor(self, inputs: TensorType) -> Tuple[np.ndarray]:
         """
-        Computes an OOD score for input samples "inputs" based on
-        the distance to nearest neighbors in the feature space of self.model
+        Compute the OOD scores for the given inputs by aggregating scores from each
+        feature layer.
+
+        The scoring process includes:
+          1. Extracting features for each layer from the input samples.
+          2. Computing the distance to the k-th nearest neighbor using the precomputed
+            FAISS index.
+          3. Aggregating per-layer scores using the specified aggregator.
 
         Args:
-            inputs: input samples to score
+            inputs (TensorType): Input samples to be scored.
 
         Returns:
-            Tuple[np.ndarray]: scores, logits
+            Tuple[np.ndarray]: A tuple containing the aggregated OOD scores.
         """
+        input_projected, _ = self.feature_extractor.predict_tensor(
+            inputs, postproc_fns=self.postproc_fns
+        )
+        scores = []
 
-        input_projected, _ = self.feature_extractor.predict_tensor(inputs)
-        input_projected = self.op.convert_to_numpy(input_projected[0])
-        input_projected = input_projected.reshape(input_projected.shape[0], -1)
-        norm_input_projected = self._l2_normalization(input_projected)
-        scores, _ = self.index.search(norm_input_projected, self.nearest)
-        return scores[:, -1]
+        for i, index in enumerate(self.indexes):
+            scores.append(self._score_layer(index, input_projected[i]))
+
+        if self.aggregator is not None:
+            aggregated_scores = self.aggregator.aggregate(scores)
+        elif len(scores) == 1:
+            aggregated_scores = scores[0]
```

Comment:
Same question as above: what if the user defines an aggregator but with a single feature_layer?

Reply:
Fixed in `FeatureBasedDetector._score_tensor` function